apps/labs/posts/scipy-array-api.md

```diff
@@ -0,0 +1,436 @@
+compiled code!
+For the functions in the extension, we can check whether the array's namespace implements the extension with
+`if hasattr(xp, 'fft')`, and use `xp.fft` if so.
+Since NumPy has an `fft` submodule, we need to be careful to still use SciPy's implementation for NumPy arrays,
+for which we have the `is_numpy` helper.
+
+Here is how that looks for `scipy.fft.fft`:
+
+```python
+def _execute_1D(func_str, pocketfft_func, x, n, axis, norm, overwrite_x, workers, plan):
+    xp = array_namespace(x)
+
+    if is_numpy(xp):
+        return pocketfft_func(x, n=n, axis=axis, norm=norm,
+                              overwrite_x=overwrite_x, workers=workers, plan=plan)
+
+    norm = _validate_fft_args(workers, plan, norm)
+    if hasattr(xp, 'fft'):
+        xp_func = getattr(xp.fft, func_str)
+        return xp_func(x, n=n, axis=axis, norm=norm)
+
+    x = np.asarray(x)
+    y = pocketfft_func(x, n=n, axis=axis, norm=norm)
+    return xp.asarray(y)
+
+def fft(x, n=None, axis=-1, norm=None,
+        overwrite_x=False, workers=None, *, plan=None):
+    return _execute_1D('fft', _pocketfft.fft, x, n=n, axis=axis, norm=norm,
+                       overwrite_x=overwrite_x, workers=workers, plan=plan)
+```
+
+After converting the production functions, we also need to convert the tests!
+
+We have the `@array_api_compatible` decorator, which parametrizes `xp` with the
+available backends that we have for testing.
+We have also written three new assertions - `assert_close`, `assert_equal` and `assert_less` -
+which incorporate checks for matching namespaces and dtypes.
+If any tests use results or input data from NumPy, we convert everything to the `xp` namespace before using
+these assertions.
+The assertions make use of functions from `xp` where possible which helps avoid device transfers.
+
+Other than that, we make similar changes to those in the production functions so that every test works
+with all of the backends we test with.
+Finally, for tests of functions which hit compiled code, we add the `@skip_if_array_api_gpu` decorator.
+
+Here is what that looks like for one of the Discrete Sine and Cosine Transforms tests:
+
+```diff
++ @skip_if_array_api_gpu
++ @array_api_compatible
+  @pytest.mark.parametrize("func", ['dct', 'dst', 'dctn', 'dstn'])
+  @pytest.mark.parametrize("type", [1, 2, 3, 4])
+  @pytest.mark.parametrize("norm", [None, 'backward', 'ortho', 'forward'])
+- def test_fftpack_equivalience(func, type, norm):
++ def test_fftpack_equivalience(func, type, norm, xp):
+      x = np.random.rand(8, 16)
++     fftpack_res = xp.asarray(getattr(fftpack, func)(x, type, norm=norm))
++     x = xp.asarray(x)
+      fft_res = getattr(fft, func)(x, type, norm=norm)
+-     fftpack_res = getattr(fftpack, func)(x, type, norm=norm)
+
+-     assert_allclose(fft_res, fftpack_res)
++     xp_assert_close(fft_res, fftpack_res)
+```
+
+## Machinery, Helpers and Utilities
```

Comment:
I planned this to be the heading for this section, although I found there wasn't too much to say under it. Maybe changing it to just 'Machinery' or something else would be better? I ended up mentioning most of the other machinery above already.

Reply:
Thoughts on removing the heading altogether and updating the previous section heading to `How to Convert a Submodule: Machinery, Helpers, and Utilities`?

Reply:
combining it all into one section seems okay, although I'm not sure I like that heading...

Reply:
Maybe it works to make it a subheading (`###`) and name it something like "Enabling and testing with non-NumPy arrays"?

Reply:
Sounds good, I've done exactly that in https://github.com/Quansight/Quansight-website/pull/770/commits/c359314a43da5b53526b6c237d02e1251c4f9eac.